src/Spice86.Core/Emulator/OperatingSystem/DosMemoryManager.cs

```diff
@@ -162,10 +182,127 @@ public DosErrorCode TryModifyBlock(in ushort blockSegment, in ushort requestedSi
+    /// allocation functions in this class. However EXE files are more complex. They require space
+    /// in memory for the PSP, the EXE itself (minus the header), its stack, and any extra memory
+    /// that the EXE may optionally request in its header. To further complicate matters, the EXE
+    /// may not request any extra memory, or it may request anywhere between a minimum and a maximum
+    /// amount. DOS is supposed to allocate as much as it can from that requested extra memory
+    /// allocation, but no less than the minimum amount. This function does that
+    /// allocation.<br/><br/>
+    /// Since determining how much memory is needed to load the EXE and what the largest block
+    /// available that can fulfill it requires knowledge of both the EXE header and the available
+    /// conventional memory space, this function is implemented in the memory manager so that it can
+    /// more easily determine the appropriate allocation for the EXE and allocate a new block of
+    /// memory for it. That also has the side-effect of making it easier to unit test the more
+    /// complicated logic of allocating the correct amount of memory without involving the process
+    /// manager or actually loading the executable into memory.
+    /// </remarks>
+    /// <param name="exeFile">EXE file header that defines the amount of space we need.</param>
+    /// <param name="pspSegment">Segment address where the PSP before the EXE will be loaded.</param>
+    /// <returns>
+    /// The <see cref="DosMemoryControlBlock"/> allocated for the program and its stack,
+    /// or <c>null</c> if no memory block was allocated.
+    /// </returns>
+    public DosMemoryControlBlock? ReserveSpaceForExe(DosExeFile exeFile, ushort pspSegment = 0) {
+        // Every program requires at least enough space for itself and the 16 paragraph (256 byte)
+        // PSP that precedes it.
+        ushort baseSizeInParagraphs = (ushort)(exeFile.ProgramSizeInParagraphs + 0x10);
+
+        ushort minSizeInParagraphs = (ushort)(baseSizeInParagraphs + exeFile.MinAlloc);
+        ushort maxSizeInParagraphs = (ushort)(baseSizeInParagraphs + exeFile.MaxAlloc);
+
+        // If both the minimum and maximum allocation fields in the EXE header are cleared, DOS will
+        // allocate the largest available block for it, and it will load the program image as high
+        // as possible in memory. We don't need to worry about loading it. That's
+        // DosProcessManager's job. We just need to make sure that we allocate the largest available
+        // block correct in this case, and that it still meets the minimum required size for the PSP
+        // and program image (our baseSizeInParagraphs). See the osdev wiki entry on the DOS EXE
+        // format (wiki.osdev.org/MZ) for more information.
+        if (exeFile.MinAlloc == 0 && exeFile.MaxAlloc == 0) {
+            ushort freeSizeInParagraphs = 0;
```

Comment:
I would move the computation of this to a new method, this could make the logic a bit smoother to understand (because return means further code is not executed)